### PR TITLE
Diff with file size

### PR DESCRIPTION
**Pull request by dstorrs**
In the base branch, robust-watch detects changes based on access time and file/directory permissions.  This PR adds file size to the list of things that are checked.  It also refactors various code to comply with requests from earlier PR.

### CODE REVIEW

**Comment by zyrolasting** [comment]
Thanks for doing this. See my other comment re: Git commands to help you deal with isolating semantic changes.

File: threads.rkt [hunk 1]
```diff
@@ -35,6 +36,9 @@
 (define (report-change . rest)
   (async-channel-put (file-activity-channel) rest))
 
+(define (report-change-literal arg)
```
Comment (zyrolasting) [inline threads.rkt:39]: `(apply report-change messages)` does the same thing as `(report-change-literal messages)`, so I'm unsure why `report-change-literal` is necessary.

**Comment by dstorrs** [comment]
On Mon, Jun 1, 2020 at 7:10 PM Sage Gerard <notifications@github.com> wrote:

> *@zyrolasting* requested changes on this pull request.
>
> Much better. There's still a trailing closing bracket and an extraneous
> whitespace change in the tests, but you can leave them alone if you want.
> I'll just do a formatting pass later.
>
> I thought about your comments in #6
> <https://github.com/zyrolasting/file-watchers/pull/6> more and better
> understand what you are trying to do. I originally assumed that you wanted
> to batch messages period, but I didn't register that you wanted to know
> that the messages came from the same poll. That concern is unique to robust
> watch, so sorry for confusing the conversation with other models. The
> intensive and apathetic watch are only ever concerned about singular files
> or directories outside of the context of a poll, so batching (as you
> describe it) doesn't make sense for them anyway.
>
> I left a couple of comments re: file existence in my review. Check on them
> and I'll do a more thorough manual test once you put in any finishing
> touches.
> ------------------------------
>
> In robust-watch.rkt
> <https://github.com/zyrolasting/file-watchers/pull/7#discussion_r433530701>
> :
>
> > @@ -25,19 +27,18 @@
>
>
>
>  (define (get-file-attributes path)
>
>    (with-handlers ([exn:fail? (λ _ #f)])
>
> -                 (cons (file-or-directory-modify-seconds path)
>
> -                       (file-or-directory-permissions path 'bits))))
>
> +    (list (file-or-directory-modify-seconds path)
>
> +          (file-or-directory-permissions path 'bits)
>
> +          (if (file-exists? path)
>
>
> This bit won't do what you're expecting. The expression is already within
> a with-handlers clause that makes the procedure return #f on any exn:fail.
> The file-or-directory-* procedures raise exn:fail:filesystem if the file
> does not exist, so the file-exists? check is redundant. It actually
> wouldn't get a chance to evaluate UNLESS the file existed in the first
> place. The literal 0 will never evaluate, and the only time you will see
> a (list _ _ 0) value is when the file exists and is empty.
>
> I see that a 0 size makes (apply + attrs) work, but there should be an
> explicit distinction between a file with no size and a file that does not
> exist. The -1 in get-listing-numbers was meant to capture that.
>

That `file-exists?` is not testing whether the path exists, it's testing
whether the thing described by that path is a file (or a symlink that
resolves to a file) as opposed to a directory (or a symlink that resolves
to a directory).

I've rewritten it to this in order to make it more obvious:

(define (get-file-attributes path)
  (with-handlers ([exn:fail? (λ _ #f)])
    (list (file-or-directory-modify-seconds path)
          (file-or-directory-permissions path 'bits)
          (cond
            [(file-exists? path)      (file-size path)] ; does path resolve
to a file?
            [(directory-exists? path) 0]                ; does path resolve
to a directory?
            [else                     0]))))


Given the above code:
1) `get-file-attributes` will return #f if the path does not exist on disk
2) If the path points to a file (i.e. not a directory) or a symlink that
resolves to a file then the last element in the return value will be the
file size
3) If the path points to a directory or a symlink that resolves to a
directory then the last element in the return value will be 0
4) If the path points to anything that answers #f to both file-exists? and
directory-exists? then the last element in the return value will be 0

The `else` clause should never be activated on a *nix / macOS machine but
is included in case Windows junctions or some other edge case returns #f to
both `file-exists?` and `directory-exists?`

Note that symlinks answer based on the nature of what they point to:

(file-exists? path-to-a-symlink-that-points-to-a-file)  ; #t
(file-exists? path-to-a-symlink-that-points-to-a-directory)  ; #f
(directory-exists? path-to-a-symlink-that-points-to-a-directory)  ; #t
(directory-exists? path-to-a-symlink-that-points-to-a-file)  ; #f


------------------------------
>
> In threads.rkt
> <https://github.com/zyrolasting/file-watchers/pull/7#discussion_r433533480>
> :
>
> > @@ -35,6 +36,9 @@
>
>  (define (report-change . rest)
>
>    (async-channel-put (file-activity-channel) rest))
>
>
>
> +(define (report-change-literal arg)
>
>
> (apply report-change messages) does the same thing as (report-change-literal
> messages), so I'm unsure why report-change-literal is necessary.
>

`report-change` was not exported -- all reporting is done by way of the
functions generated from `report-iface`.  Those functions modify their
argument by prepending the type of the watch (in this case 'robust), which
does not work when returning a list of events, each of which should have
the type.  `report-change-literal` was intended to be a reporting function
that said "report exactly what I'm giving you" without reference to whether
it's one event, a list of events, or etc.

**Comment by zyrolasting** [comment]
Tests pass and docs look fine. Anything else I would change is more stylistic than functional, so I'll approve this here. Thank you again for your contribution! I'll add you to a credits section in the README. James too? I don't see his username anywhere.

**Pull request merged**